app/assets/javascripts/wpcc/components/Email.js

```diff
@@ -0,0 +1,88 @@
+        contributionsHeading: 'Your contributions',
+        resultsHeading: 'Your Results',
+        qualifyingEarningsHeading: 'Qualifying earnings'
+      }
+    };
+
+  Email = function($el, config) {
+    Email.baseConstructor.call(this, $el, config, defaultConfig);
+
+    this.$emailLink = this.$el.find('[data-dough-email-link]');
+    this.$detailsHeadingSummary = this.$el.find('[data-dough-details-heading-summary]');
+    this.$contributionsHeadingSummary = this.$el.find('[data-dough-contributions-heading-summary]');
+    this.$eligibleSalary = this.$el.find('[data-dough-eligible-salary]');
+    this.$resultsTables = this.$el.find('[data-dough-results-table]');
+    this.$frequencySelector = this.$el.find('[data-dough-selector]');
+  };
+
+  /**
+  * Inherit from base module, for shared methods and interface
+  */
+  DoughBaseComponent.extend(Email);
+
+  Email.componentName = 'Email';
+
+  Email.prototype.init = function(initialised) {
+    if (this.$emailLink.length > 0) {
+      this._initialisedSuccess(initialised);
+      this._setUpEvents();
+    }
+  };
+
+  Email.prototype._setUpEvents = function() {
+    var _this = this;
+
+    this._setUpLink();
+    this.$frequencySelector.change(function() {
+      _this._setUpLink();
+    });
+  }
+
+  Email.prototype._setUpLink = function() {
+    var message = this._getMessage();
+    var href = 'mailto:?body=' + message;
+
+    this.$emailLink.prop('href', href);
+  }
+
+  Email.prototype._getMessage = function() {
+    var message = '';
+    var i18nStrings = this.config.i18nStrings;
+
+    message += '1. ' + i18nStrings.detailsHeading + ': ' + this.$detailsHeadingSummary.text().trim() + '\n\n';
+    message += '2. ' + i18nStrings.contributionsHeading + ': ' + this.$contributionsHeadingSummary.text().trim() + '\n\n';
+    message += '3. ' + i18nStrings.resultsHeading + '\n';
+    message += i18nStrings.qualifyingEarningsHeading + ': ' + this.$eligibleSalary.text().trim() + '\n\n';
+
+    for (var i = 0, max = this.$resultsTables.length; i < max; i++) {
+      var resultTable = this.$resultsTables[i];
+
+      message += $(resultTable).find('[data-dough-results-period-title]').text().trim() + '\n';
+      message +=
+        $(resultTable).find('[data-dough-period-heading-yours]').text().trim() + ': ' +
+        $(resultTable).find('[data-dough-employee-contribution]').text().trim() + ' ' +
+        $(resultTable).find('[data-dough-tax-relief]').text().trim() + '\n';
+      message +=
+        $(resultTable).find('[data-dough-period-heading-employers]').text().trim() + ': ' +
+        $(resultTable).find('[data-dough-employer-contribution]').text().trim() + '\n';
+      message +=
+        $(resultTable).find('[data-dough-period-heading-total]').text().trim() + ': ' +
+        $(resultTable).find('[data-dough-total]').text().trim();
+
+      if (i !== max - 1) {
+        message += '\n\n';
+      }
+    }
+
+    return encodeURIComponent(message);
+  }
+
+  return Email;
+});
```

Comment:
Can this be dried up in any way?

Reply:
Not sure what this means.

Reply:
alot of the processing appears to be similar e.g. 
1. get some text
2. trim it
3. add some newlines

the headings look similar - maybe it should be in a separate method, just to make more readable

I dont know how string interpolation in js works, but can you do something like

``` 
var headings = getHeadings();
message += headings;

Email.prototype._getHeadings = function() {
  var headings = ['details', 'contributions', 'results'];
  var headingsText = "";

  for (var i = 0, max = this.headings.length; i < max; i++) {
    var headingKey = headings[i] + "Heading";
    var summary = $headings[i] + "HeadingSummary"
    headingsText += (i++) + '. ' + i18nStrings.headingKey + ': ' + this.summary.text().trim() + '\n\n';
  }
  return headingsText;
}```

Reply:
Get your point but:

- I'm not sure that would make this more significantly more readable for just three elements (with slight differences between them, such as one or two new lines which would complicate the new method a bit more) 
- I want to mirror the structure of spec/javascripts/tests/Email_spec.js as far as possible to aid with potential bug fixing

So on balance I'd rather leave it as it is.

Reply:
It would make sense to put the heading code, even if as is, in a separate function to make more readable. I would also argue that we should not structure code to fit the tests. The tests should be structured to test the code, although I do understand the temptation.

Reply:
Yes, I get the point about the test/code relationship; this reflects how it evolved really: the test produced the desired output then the code built that dynamically.

Reply:
I think this is ok as it is, the code is readable and it works